Fix crash on instrument viewer draw tab

DannyHindson opened the pull request:
**Description of work.**

Some of the items on the "Apply and Save" drop down on the Draw tab calls GroupDetectors. This can fail for certain types of workspace (eg if workspace doesn't have common bins). This was previously causing workbench to crash so I've made a change to catch any exceptions from the GroupDetectors call and report the error to the user.

**To test:**

See linked issue

Fixes #33232.


---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DannyHindson committed:
Catch errors from calls to GroupDetectors and report to user

Harrietbrown reviewed (changes requested):
This has fixed the crash, is it possible to write a test for this fix?

DannyHindson committed:
Add tests for save actions on the Draw tab

These involve calling some protected slots on the InstrumentWidgetMaskTab
class. It's possible to access and call these from the test using
QMetaObject::invokeMethod

Harrietbrown reviewed (approved):
Good job with the test and fixing the crash.

DavidFair commented on qt/widgets/instrumentview/test/InstrumentWidget/InstrumentWidgetTest.h:
```diff
@@ -262,26 +306,37 @@ class InstrumentWidgetTest : public CxxTest::TestSuite {
+    explicit MockInstrumentWidgetMaskTab(InstrumentWidget *instrWidget) : InstrumentWidgetMaskTab(instrWidget){};
+
+    GNU_DIAG_OFF_SUGGEST_OVERRIDE
+    MOCK_METHOD1(showMessageBox, void(const QString &));
```
Since this is a new mock can we use the new mocking decelerations that doesn't need warning suppressions

```c++
MOCK_METHOD(void, showMessageBox, (const QString &), (override))
```

This allows us to drop warning suppressions, and will catch cases where the signature changes but this mock doesn't (currently this will not produce a warning, only a runtime error)

DavidFair commented on qt/widgets/instrumentview/test/InstrumentWidget/InstrumentWidgetTest.h:
```diff
@@ -262,26 +306,37 @@ class InstrumentWidgetTest : public CxxTest::TestSuite {
-  InstrumentWidget construct(std::unique_ptr<DisplayMock> displayMock, QtMock *qtMock, GLMock *glMock,
-                             std::unique_ptr<ConnectMock> connectMock,
-                             std::unique_ptr<MetaObjectMock> metaObjectMock) const {
+  class MockInstrumentWidgetMaskTab final : public InstrumentWidgetMaskTab {
```
This should move into it's own file too

DavidFair commented on qt/widgets/instrumentview/test/InstrumentWidget/InstrumentWidgetTest.h:
```diff
@@ -205,6 +215,40 @@ class InstrumentWidgetTest : public CxxTest::TestSuite {
+    // create a second workspace that doesn't have common bins
+    auto ws_d = WorkspaceCreationHelper::create2DWorkspaceWithFullInstrument(2, 2);
+    AnalysisDataService::Instance().addOrReplace("test_ws_d", ws_d);
+    FrameworkManager::Instance().exec("ConvertUnits", 6, "InputWorkspace", "test_ws_d", "OutputWorkspace", "test_ws_d",
```
I'm not keen on this, we should create the method / exec it directly:
```c++
auto alg = ConvertUnits();
alg.initialize()
alg.setRethrows(true);
alg.setProperty("InputWorkspace", "test_ws_d")
alg.execute()
//etc.
```

DavidFair reviewed (changes requested):
Some modern patterns / coding standards changes

I'm with these changes otherwise

DavidFair commented:
Haha Gemma beat me by one minute :laughing: 
Feel free to do a follow up PR on `main`, I don't feel super strongly enough to revert / complain :slightly_smiling_face:

DannyHindson commented:
> Haha Gemma beat me by one minute 😆 Feel free to do a follow up PR on `main`, I don't feel super strongly enough to revert / complain 🙂

No problem. I'll see if I find time today to do a follow up PR. I noticed some places used MOCK_METHOD and others MOCK_METHOD1 but didn't know why so that's good to know thanks
